### ISSUE TITLE
[GSoC 2019] Growing the model garden 🌳

### ISSUE THREAD

**Issue by sdll** [tensorflow/tfjs#1684]
This issue tracked the progress for GSoC'19 project [**"Reasonable Effectiveness of Mobile Inference: Adaptive Growth of the TensorFlow.js Model Garden"**](https://summerofcode.withgoogle.com/projects/#5703734951149568).

The final report is [here](https://sdll.io/tfjs-model-garden).

## Active

- [x] **DeepLab**

  * Status: **Merged into the model garden.**
  * Requires compute: no
  * [PR](https://github.com/tensorflow/tfjs-models/pull/229)
  * [Reference implementation](https://github.com/tensorflow/models/tree/master/research/deeplab)

- [x] **EfficientNet**

  * Status: **The port was completed, but the model did not provide 10x improvements over MobileNet and will be moved to the [personal repo](https://github.com/sdll/efficientnet-js).**
  * Requires compute: no
  * [PR](https://github.com/tensorflow/tfjs-models/pull/238)
  * [Reference implementation](https://github.com/qubvel/efficientnet)

- [ ] **Text Detection**
  * Status: **The js code is under review. The optimized model requires more training to pass the quality tests for the merge.**
  * Requires compute: yes
  * Reference implementations: ([tf.slim](https://github.com/liuheng92/tensorflow_PSENet)) ([PyTorch](https://github.com/whai362/PSENet))
  * [PR](https://github.com/tensorflow/tfjs-models/pull/274)

- [ ] **Text Recognition**
  * [Reference implementation](https://github.com/bgshih/aster)
  * Status: **The port from pytorch to tf.keras is [under development](https://github.com/sdll/moran).**
  * Requires compute: yes

## Scratchpad
- [ ] **VGGish**
  * Status: **Not in active development**
  * Requires compute: no
  * [Reference implementation](https://github.com/tensorflow/models/tree/master/research/audioset)
  * [Usage Example](https://resources.wolframcloud.com/NeuralNetRepository/resources/VGGish-Feature-Extractor-Trained-on-YouTube-Data)

- [ ] **DeepGaze**

  * Status: **Not in active development**
  * Requires compute: yes
  * [Reference implementation](https://github.com/mpatacchiola/deepgaze/tree/2.0)
  * Motivation: [lookie-lookie](https://cpury.github.io/learning-where-you-are-looking-at/)

- [ ] **Style Transfer**

  * Status: **Waiting for the core team feedback**
  * Requires compute: no 
  * Existing implementations: [(1)](https://github.com/reiinakano/arbitrary-image-stylization-tfjs) ([description & lines of improvement](https://magenta.tensorflow.org/blog/2018/12/20/style-transfer-js/))
- [ ] **Pix2Pix**

  * Status: **Waiting for the core team feedback**
  * Requires compute: no 
  * [Reference implementation](https://github.com/affinelayer/pix2pix-tensorflow)
  * Existing TF.js ports: [(1)](https://github.com/yining1023/pix2pix_tensorflowjs)

- [ ] **Speech to Text**
  * Status: **Not in active development**
  * Requires compute: no
  * [Reference implementation](https://github.com/mozilla/DeepSpeech)
  * [Motivation](https://github.com/tensorflow/tfjs/issues/1725)
- [ ] **Face Swap**
  * Status: **Not in active development**
  * Requires compute: possibly
  * [Reference implementation](https://github.com/YadiraF/PRNet)
  * Existing TF.js ports: [(1)](https://github.com/kylemcdonald/prnetjs)

- [ ] **Sketch Simplification**

  * Status: **Not in active development, awaiting the decision on licensing**
  * Requires compute: no
  * [Reference implementation](https://github.com/bobbens/sketch_simplification)
- [ ] **DeOldify**

  * Status: **Not in active development**
  * Requires compute: yes 
  * [Reference implementation](https://github.com/jantic/DeOldify)

- [ ] **HNAtt**

  * Status:
	- **Waiting for the core team feedback**
  * Requires compute: yes
  * [Reference implementation](https://github.com/minqi/hnatt)

- [ ] **VAMPIRE**
  * Status: **Not in active development**
  * Requires compute: yes
  * [Reference implementation](https://github.com/allenai/vampire)
  * Motivation: Might be a good addition either to tfjs-models or tfjs-examples as an alternative to USE for domain-specific text analysis.
- [ ] **CycleGAN**

  * Status: **Not in active development**
  * Requires compute: yes 
  * [Reference implementation](https://github.com/LynnHo/CycleGAN-Tensorflow-2)

- [ ] **EDANet**
  * Status: **Not in active development**
  * Requires compute: no
  * [Reference implementation](https://github.com/wpf535236337/pytorch_EDANet)
- [ ] **ShuffleNet**
  * Status: **Not in active development**
  * Requires compute: yes 
  * [Reference implementation](https://github.com/MG2033/ShuffleNet)
- [ ] **ErfNet**
  * Status: **Not in active development**
  * Requires compute: yes 
  * [Reference implementation](https://github.com/ronrest/erfnet_segmentation)
- [ ] **YOLO**

  * Status: **Not in active development**
  * Existing TF.js port: [(1)](https://github.com/ModelDepot/tfjs-yolo-tiny) 
  * Requires compute: no
  * [Reference implementation](https://github.com/wizyoung/YOLOv3_TensorFlow)

- [ ] **Mask-RCNN**

  * Status: **Not in active development**
  * Requires compute: yes 
  * [Reference implementation](https://github.com/matterport/Mask_RCNN/)

**Comment by sdll** [comment]
@nsthorat (as I have promised...), @dsmilkov, @manrajgrover, let me know if you have any opinion on which models you would like to see first. I would love to hear any ideas and high-level views on the details of implementation as well.

**Comment by kavikode** [comment]
Mask-RCNN and YOLO will be very helpful to see first @sdll @nsthorat @dsmilkov @manrajgrover because they are very useful and commonly used for computer vision.

**Comment by sdll** [comment]
@kavikode, TF.js implementations of YOLO already [exist](https://github.com/ModelDepot/tfjs-yolo-tiny). To convert Mask-RCNN as described [here](https://github.com/tensorflow/tfjs/issues/1094), you might try the following:

- switch the source to tf.keras
- load the pre-trained weights
- export the model to tf.SavedModel
- convert tf.SavedModel using tfjs-converter

However, Mask-RCNN is quite a heavyweight model and may require further optimizations to run smoothly.

**Comment by kavikode** [comment]
thank you so much @sdll

**Comment by beriberikix** [comment]
What is the process to pitch for moving items from the `scratchpad` to the `active` list? I reported #1725 and love to make a case to whoever will listen :)

**Comment by sdll** [comment]
@beriberikix, thank you for your interest! I have not dived into the details, but DeepSpeech seems to be resource-intensive, given the discussion in your [other issue](https://github.com/mozilla/DeepSpeech/issues/2233). Have you had any success converting the model using `tfjs-converter`, how does it fare?

**Comment by beriberikix** [comment]
I haven't had any success with the converter tool. I've asked for some clarification on how the saved model is generated. I'll updated if and when I get the converter working.

**Comment by jessetrana** [comment]
I've worked some with DeepSpeech in the past. As I recall, it has both significant compute work for generating the input audio features (a la MFCC) as well as decoding via the language model. I'm not sure how much of that functionality also has corresponding JS libraries, so it's possible that porting it may be an involved effort. By any chance, have you looked into the supporting work apart from the TF model yet to see how easy or hard it might be?

**Comment by sdll** [comment]
Thanks everyone for the suggestions. Since the GSoC 2019 run came to the end, I have added the final report to the issue and updated the status of each port. Let me know if you have any other ideas on porting text detection/text recognition or would like to work on other models.

**Comment by gautamchitnis** [comment]
> @kavikode, TF.js implementations of YOLO already [exist](https://github.com/ModelDepot/tfjs-yolo-tiny). To convert Mask-RCNN as described [here](https://github.com/tensorflow/tfjs/issues/1094), you might try the following:
> 
> * switch the source to tf.keras
> * load the pre-trained weights
> * export the model to tf.SavedModel
> * convert tf.SavedModel using tfjs-converter
> 
> However, Mask-RCNN is quite a heavyweight model and may require further optimizations to run smoothly.

Hey

I'm looking to implement MaskRCNN inference with tfjs for a production project I'm working on. I would love to get some help and would very much like to contribute for getting MaskRCNN into model garden if possible.

Please let me know where/how we can discuss this and take it forward.

From what I have achieved till now, it looks like following things are required for my particular case:

Rewrite ```mold_inputs()``` and ```get_anchors()``` functions similar to their python implementation over at [this repo].(https://github.com/bendangnuksung/mrcnn_serving_ready/blob/master/inferencing/saved_model_preprocess.py)

**Comment by babanthierry94** [comment]
Hi @gautamchitnis @kavikode 
Please i'm looking for a way to implement mask rcnn inference on a web app. Please can you help me. I try to use tensorflowjs.
But have got an error while loading de model.json file.
